### DISCUSSION TITLE
Integrating Spectrum Web Components into an Adobe CEP Panel

### DISCUSSION THREAD

**Discussion by tpfleging** [adobe/spectrum-web-components#2423] [General]
I'm relatively new to programming (<1 year) so apologies on any dumb questions. Would love some advice from any experts on here like @Westbrook 

For context I'm trying to use Spectrum Web Components to build out an Adobe Premiere CEP Panel.  Two issues I'm having:
1) I can get most everything to work using the unrecommended setup using src="https://jspm.dev/@spectrum-web-components/bundle/elements.js" in the HTML.  However, text fields and number fields don't let the user input any values.  Maybe this is an issue with the unrecommended setup?

2) I'm having trouble figuring out the recommended set up of using individual components with LitElement.  Do I have to use Typescript for this or can I use just JS and HTML?  I've never used TypeScript and it seems more rare to use with Adobe CEP.

Thanks!!

**Comment by Westbrook** [comment]
I don't really know much of the specifics for integrating with Adobe Premiere CEP Panel, but let's see what we can learn here together!

1. When you say "don't let the use input any values" do you literally mean that they can't type into the fields? or something else? If you were able to share a video of the experience you are having here, it may go a long way to clarifying what to are seeing for the team here. Additionally, you may want to track https://github.com/adobe/spectrum-web-components/issues/2369 where we are looking at adding better support for use of our elements within `<form>` elements.
2. You can absolutely leverage our package with HTML and JS. We develop them in TS, but DO NOT ship TS to users. All consumption can happen in the language that _you_ the consuming developer feel most comfortable with.

**Reply by tpfleging** [reply]
Amazing thanks so much for the help!!

1) Attached is an example of what I mean for that.  The textfield or numberfield visually look fine but when I'm typing it just doesn't let me type anything in.  Picker and checkbox both work fully functionally but these don't for some reason.
https://user-images.githubusercontent.com/109366868/179277273-59cb3961-e105-462d-87b2-8d493d7e14a4.mp4

2) Awesome! So the yarn I run in the command line and then these other two lines I would include in my JS?
<img width="831" alt="Screen Shot 2022-07-15 at 1 31 06 PM" src="https://user-images.githubusercontent.com/109366868/179277641-6a6a48e5-6cf4-4027-8240-d5ae92683bc8.png">

**Reply by Westbrook** [reply]
For most things, you only need the registering import:
```
import '@spectrum-web-components/number-field/sp-number-field.js';
```
Which will ensure that the component is available in your page.

The experience in the recording is super weird. I wonder if there is something about Adobe Premiere CEP Panel that is doing something weird there? Are you able to reach out to support on their end to see what might be happening?

**Reply by tpfleging** [reply]
Thanks!! I posted on the Premiere SDK forums so I'll see if they have any input on that.

Question on the registering import because I can't seem to get that to work.  Hopefully last question for you because I know it's been a lot.  I have this setup:

A JS file with just this
```
import '../spectrum-web-components/packages/theme/sp-theme.js';
import '../spectrum-web-components/packages/theme/src/themes.js';
import '../spectrum-web-components/packages/number-field/sp-number-field.js';
```

An HTML file with this
```
<!doctype html>
<html>
<head>
<meta charset="utf-8">
<script src="js/importSWC.js"></script>
</head>
<body scroll="yes">
	<div class='center'>
	
	<sp-theme scale="medium" color="darkest">
	<sp-number-field value="1024" style="width: 200px"></sp-number-field>
	</sp-theme>

</div>
</body>
</html>
```

But this doesn't seem to work.  Am I missing something?

**Reply by tpfleging** [reply]
Update:  I found a workaround for the text field / number field issue by adding in contentEditable="true" to each instance in the HTML.  Can't use the number stepper still but that's fine.

Open issue still:  I can't get the script src in the HTML to reference the node modules locally instead of the jspm.dev website.  It's not a Premiere CEP issue because it doesn't show up correctly in my IDE HTML preview either.  But everything works correctly when I use the jspm.dev website.

**Reply by Westbrook** [reply]
`contentEditable` does some other things that I'm not sure you mean, and is specifically the sort of workaround that I would think comes with peculiarities of Premiere CEP. You should definitely see if they can't get you information about that.

For the imports, you _should_ be able to use:
```
import '@spectrum-web-components/theme/sp-theme.js';
```
But, if you don't have a bundler that resolves that, then it's hard for me to explain how you'd leverage our packages. If you HAVE to use relative pathing, then you might be able to do:
```
import '../node_modules/spectrum-web-components/theme/sp-theme.js';
```
But that opens you up to so many troubles down the line.

How are you preparing your code for delivery to production? Webpack, Rollup, etc...

**Reply by tpfleging** [reply]
Ohhh wow I think that's my issue then is that I'm not using any bundler.  This is the only module I have in the project so I didn't need one, and I haven't used one before.  Do you have a recommendation on Webpack vs Rollup?

**Reply by tpfleging** [reply]
Good news:  Webpack worked perfectly and everything shows correctly now in an HTML preview
Bad news:  Still doesn't work in a Premiere CEP Panel.  Very strange that the jspm.dev way works but this doesn't.

**Reply by Westbrook** [reply]
This is likely due to the process by which you publish to CEP. Make sure that the JS you are building with Webpack is being included in your publication and that the directory relationship between your HTML and JS is the same locally as it in the published location.

**Reply by tpfleging** [reply]
Hmmmm yea I'm not publishing it anywhere, CEP just opens it directly from that folder I'm building in.  I'm still trying to figure it out but no luck so far.
